### DISCUSSION TITLE
Handling data structures from Smart Meter Interface in script

### DISCUSSION THREAD

**Discussion by ByteJunk** [arendst/Tasmota#20342] [Support request] [answered]
Hi! I asked the Discord but was suggested that here would be a better fit.

I'm trying to pull a lot of data from a power meter using the Smart Meter Interface on a ESP8266EX; I'm struggling a bit with the scripting interface - specifically, with custom data structures.
One of these structures is 12 bytes long, and it represents a custom datetime. Any thoughts on how I can approach this, short of breaking it into a bunch of separate variables?
I'm using Modbus, not sure how I could squeeze such a struct through the "decoder".

**Comment by pkkrusty** [comment]
Struct details?

**Reply by ByteJunk** [reply]
Of course, this is the format:

2 Bytes		Long unsigned	
1 Byte		Unsigned	
1 Byte		Unsigned
1 Byte		Unsigned
1 Byte		Unsigned
1 Byte		Unsigned
1 Byte		Unsigned
1 Byte		Unsigned
2 Bytes		Long
1 Byte		Unsigned

**Reply by pkkrusty** [reply]
If the registers are all sequential, you can pull all 12 bytes (maybe you mean registers? Not sure a 2-byte long exists) at once, then break them out.  I think you'll have to do some re-arranging unless you were pulling a long int that could translate directly to epoch.

What's your meter descriptor look like? just pull all 12 registers at once with `r0103<register>000C` then break it out like
```
1,010318UUuuUUuu@i0:1,First Value,X,First_Value,0
1,010318x4UUuu@i1:1,Second Value,X,Second_Value,0
1,010318x6UUuu@i2:1,Third Value,X,Third_Value,0
etc...
```
Then assign those meter lines to variables in >S section, like `v1=sml[1] ` (requires building with USE_SML_SCRIPT_CMD)

If you really do mean 12 bytes, then pull all six registers with `r0103<register>0006` and then your descriptor lines would look more like 
```
1,01030CUUuu@i0:1,First Value,X,First_Value,0
1,01030Cx2uu@i1:1,Second Value,X,Second_Value,0
1,01030Cx3uu@i2:1,Third Value,X,Third_Value,0
etc...
```

**Reply by ByteJunk** [reply]
The 12 bytes come from a single register, a sample metric definition would be like: r010400010001
I can get it to work by slicing that up into individual metrics, like:
```
1,01040Cx05uu@i0:1,Clock,h,CH,0
1,01040Cx06uu@i0:1,Clock,m,CM,0
1,01040Cx07uu@i0:1,Clock,s,CS,0
```
But to capture the full datetime I need to use 10 variables, and there are multiple datetimes.

I'm still very unfamiliar with the scripting language, any insights are appreciated. 

I don't have to process this with the script. Perhaps if I can dump it as string or something to mqtt, I can pick it up elsewhere and process it there where I'm not so limited?

**Reply by pkkrusty** [reply, marked as answer]
I think it would be easiest to process it internally and then send a JSON datetime as text string. Or maybe convert to epoch internally. 

Assign all 10 variables like 

month=sml[1]
day=sml[2]
Etc then do your calculations

If you want specific help with the code, would need the specifics of the registers

**Reply by ByteJunk** [reply]
To sum up, there's no way to define data formats for the SML decoder, and even there was, I read somewhere that max var size is 4 bytes, so something like this isn't possible directly.

The approach is to decode parts of the chunk, and then build a JSON string.

I think I got it right... Thank you @pkkrusty for all your help! Marking this as solved.

**Comment by GEAG** [comment]
To read an 8-byte long string into var SID, I use only 2 bcd patterns and reconstruct the SID in the code:

```
>D
SID=""

>M 1
;...
; SM-ID (UI-hidden)
1,02130908bcd8@1,*,,,0
1,02130908x3xxbcd8@1,*,,,0

>S
; reconstruct SID once
if (SID=="" and smls[-1]!="")
{
 SID+=sb(smls[-1] 7 1)
 SID+=sb(smls[-1] 5 1)
 SID+=sb(smls[-1] 3 1)
 SID+=sb(smls[-1] 1 1)

 SID+=sb(smls[-2] 7 1)
 SID+=sb(smls[-2] 5 1)
 SID+=sb(smls[-2] 3 1)
 SID+=sb(smls[-2] 1 1)
}
print %SID%

```
